src/plan/barriers.rs

```diff
@@ -45,6 +46,9 @@ impl BarrierSelector {
 pub trait Barrier<VM: VMBinding>: 'static + Send + Downcast {
     fn flush(&mut self) {}
 
+    /// load referent from java.lang.Reference
+    fn load_reference(&mut self, _referent: ObjectReference) {}
```

Comment:
Java-specific stuff such as `java.lang.ref.Reference` shall eventually be moved to the `mmtk-openjdk` binding.  What mmtk-core cares about is that loading from a weak reference cell (including `java.lang.ref.Reference.get()` and also weak tables in CRuby) upgrades the weak reference to a strong reference.  The referent object was previously not strongly reachable from roots, but is now strongly reachable from the stack.  The mmtk-core should simply remember the observed referent, and make it (and its children) part of the snapshot.

Reply:
This is technically part of the barrier API since it's a read/load barrier. I don't know if we want to push that part of the barrier code into the binding. We can rename it to `load_weak_reference` if that is more descriptive if you have an issue with this clashing with `ObjectReference`.

Reply:
For SATB, we wouldn't require read/load barriers for normal strong reference fields.  We only need it when loading from a weak reference field.  I think it is worth emphasizing that the `load_weak_reference` API is only for weak references.  Otherwise, users (including developers of other VM bindings) may think the SATB barrier needs read barriers for all reference fields.

Reply:
Right. That's why maybe we can rename it to be something like `load_weak_reference` (or `load_from_weak_reference`).